README.md
# Platform Deployment Manager

* [Design](#design)
* [Building](#building)
* [API Documentation](#api-documentation)
  * [Repository API](#repository-api)
  * [Packages API](#packages-api)
  * [Applications API](#applications-api)
  * [Environment Endpoints API](#environment-endpoints-api)
* [Deployment Manager Variables](#deployment-manager-variables)

# Design #

The Deployment Manager is a service that manages package deployment and application creation for a single PNDA cluster.

- It implements the Packages, Applications, Repository and EnvironmentEndpoints API used by operators.
- It parses and validates basic package structure.
- It interacts with a Repository and a Registrar to determine available & record currently deployed packages and applications.
- It includes a number of component specific Creator implementations that carry out the concrete steps necessary to set up different parts of the Core Platform.
- It is easily extensible to support additional component types and repository types.

The design consists of a main class that implements the APIs and coordinates between a Repository, a Registrar and an Application Creator that dynamically loads a number of component specific Creator classes as required by a particular package.

HTTP and Python bindings are provided for these APIs.

## Connecting

By default, the Deployment Manager is installed on `edge` node o. In order to access it go to: http://[cluster-name]-cdh-edge:5000

## Repository ##

Packages are made available via a repository. The Deployment Manager is configured with a client of this repository at instantiation time. The reference repository is implemented as a thin wrapper over an Openstack Swift container.

## Registrar ##

The details of package deployments for a given service instance are recorded by a registrar. The registrar stores information in HBase in the platform_packages and platform_applications tables.

## Application Creator ##

The Application Creator handles the creation and control of applications on behalf of the Deployment Manager. It implements business logic that is common to all components and delegates to a component specific Creator as required by a particular package. Creator subclasses are dynamically loaded as needed by the Application Creator.

## Creator ##

Each component type is associated with a subclass of Creator. Each Creator implements the specific steps necessary to perform the following functions: 

### Validation ###

Each component type has a specific structure. Each Creator implements a validation function that checks that structure. All components are validated before the package is deployed. If any validation function fails, the package is deemed “bad” and package deployment fails. This provides an opportunity to catch simple package construction problems early in the deployment process.

### Application creation ###

Each component type has specific creation requirements and resource dependencies. Each Creator implements the process required to create components of a given type and returns “application_data”. The Deployment Manager aggregates the application data generated by the process of creating each of the components in the package, then persists an association between this and the package deployment using a Registrar.

### Application control ###

Applications may be paused and restarted. This leaves all the installed components in-place and temporarily stops the running processes associated with those components.

### Undeployment ###

Each Creator implements a specific set of steps to uninstall components of its associated type. The Creator is passed the application data associated with the package and component and uses this to execute those steps.


# Requirements

* [Maven](https://maven.apache.org/install.html)

# Building

To build the Deployment Manager, change to the `api` directory, which contains the `pom.xml` file. Type `mvn clean package` on the command line. Once the build is successful, the built package will be placed in the `target` folder.

# API Documentation

* [Repository API](#repository-api)
  * [GET /repository/packages](#list-packages-from-the-repository)
* [Packages API](#packages-api)
  * [GET /packages](#list-packages-currently-deployed-to-the-cluster)
  * [GET /packages/_package_/status](#get-the-status-for-package)
  * [GET /packages/_package_](#get-full-information-for-package)
  * [PUT /packages/_package_](#deploy-package-to-the-cluster)
  * [DELETE /packages/_package_](#undeploy-package-from-the-cluster)
* [Applications API](#applications-api)
  * [GET /applications](#list-all-applications)
  * [GET /packages/_package_/applications](#list-applications-that-have-been-created-from-package)
  * [GET /applications/_application_/status](#get-the-status-for-application)
  * [POST /applications/_application_/start](#start-application)
  * [POST /applications/_application_/stop](#stop-application)
  * [GET /applications/_application_](#get-full-information-for-application)
  * [PUT /applications/_application_](#create-application-from-package)
  * [DELETE /applications/_application_](#destroy-application)
* [Environment Endpoints API](#environment-endpoints-api)
  * [GET /environment/endpoints](#list-environment-variables-known-to-the-deployment-manager)

## Repository API

### List packages from the repository

?recency=n may be used to control how many versions of each package are listed, by default recency=1
````
GET /repository/packages

Response Codes:
200 - OK
500 - Server Error

Example response:
[
    {
	"latest_versions": [{
		"version": "1.0.23",
		"file": "spark-batch-example-app-1.0.23.tar.gz"
	}],
	"name": "spark-batch-example-app"
    }
]
````

## Packages API

### List packages currently deployed to the cluster
````    
GET /packages

Response Codes:
200 - OK
500 - Server Error

Example response:
["spark-batch-example-app-1.0.23"]
````

### Get the status for _package_
````
GET /packages/<package>/status

Response Codes:
200 - OK
500 - Server Error

Example response:
{"status": "DEPLOYED", "information": "human readable error message or other information about this status"}

Possible values for status:
NOTDEPLOYED
DEPLOYING
DEPLOYED
UNDEPLOYING
````

### Get full information for _package_
````
GET /packages/<package>

Response Codes:
200 - OK
500 - Server Error

Example response:
{
	"status": "DEPLOYED",
	"version": "1.0.23",
	"name": "spark-batch-example-app",
	"defaults": {
		"oozie": {
			"example": {
				"end": "${deployment_end}",
				"start": "${deployment_start}",
				"driver_mem": "256M",
				"input_data": "/user/pnda/PNDA_datasets/datasets/source=test-src/year=*",
				"executors_num": "2",
				"executors_mem": "256M",
				"freq_in_mins": "180",
				"job_name": "batch_example"
			}
		}
	}
}
````

### Deploy _package_ to the cluster
````
PUT /packages/<package>

Response Codes:
202 - Accepted, poll /packages/<package>/status for status
404 - Package not found in repository
409 - Package already deployed
500 - Server Error
````

### Undeploy _package_ from the cluster
````
DELETE /packages/<package>

Response Codes:
202 - Accepted, poll /packages/<package>/status for status
404 - Package not deployed
500 - Server Error
````

## Applications API

### List all applications
````
GET /applications

Response Codes:
200 - OK
500 - Server Error

Example response:
["spark-batch-example-app-instance"]
````

### List applications that have been created from _package_
````
GET /packages/<package>/applications

Response Codes:
200 - OK
500 - Server Error

Example response:
["spark-batch-example-app-instance"]
````

### Get the status for _application_
````
GET /applications/<application>/status

Response Codes:
200 - OK
404 - Application not known
500 - Server Error

Example response:
{"status": "STARTED", "information": "human readible error message or other information about this status"}

Possible values for status:
NOTCREATED
CREATING
CREATED
STARTING
STARTED
STOPPING
DESTROYING
````

### Get run-time details for _application_
````
GET /applications/<application>/detail

Response Codes:
200 - OK
404 - Application not known
500 - Server Error

{
        "status": "STARTED",
        "name": "spark-batch-example-app-instance",
        "yarn-ids": [
            {"component":"example", "type":"oozie", "yarn-id":"application_1455877292606_0404"}
        ]
}
````

### Start _application_
````
POST /applications/<application>/start

Response Codes:
202 - Accepted, poll /applications/<application>/status for status
404 - Application not known
500 - Server Error
````

### Stop _application_
````
POST /applications/<application>/stop

Response Codes:
202 - Accepted, poll /applications/<application>/status for status
404 - Application not known
500 - Server Error
````

### Get full information for _application_
````
GET /applications/<application>

Response Codes:
200 - OK
404 - Application not known
500 - Server Error

Example response:
{
	"status": "CREATED",
	"overrides": {
		"oozie": {
			"example": {
				"executors_num": "5"
			}
		}
	},
	"package_name": "spark-batch-example-app-1.0.23",
	"name": "spark-batch-example-app-instance",
	"defaults": {
		"oozie": {
			"example": {
				"end": "${deployment_end}",
				"input_data": "/user/pnda/PNDA_datasets/datasets/source=test-src/year=*",
				"driver_mem": "256M",
				"start": "${deployment_start}",
				"executors_num": "2",
				"freq_in_mins": "180",
				"executors_mem": "256M",
				"job_name": "batch_example"
			}
		}
	}
}
````

### Create _application_ from _package_

````
PUT /applications/<application>
{
	"package": "<package>",
	"<componentType>": {
		"<componentName>": {
			"<property>": "<value>"
		}
	}
}

Response Codes:
202 - Accepted, poll /applications/<application>/status for status
400 - Request body failed validation
404 - Package not found
409 - Application already exists
500 - Server Error

Example body:
{
	"package": "<package>",
	"oozie": {
		"example": {
			"executors_num": "5"
		}
	}
}

Package is mandatory, property settings are optional
````

### Destroy _application_
````
DELETE /applications/<application>

Response Codes:
200 - OK
404 - Application not known
500 - Server Error
````

## Environment Endpoints API
### List environment variables known to the deployment manager
````
GET /environment/endpoints

Response Codes:
200 - OK
500 - Server Error

Example response:
{"zookeeper_port": "2181", "cluster_root_user": "cloud-user", ... }
````
# Deployment Manager Variables #

The following variables are made available for use in the configuration files for every component and injected as previously described.

## Component Variables ##
````
component_application   unique application ID
component_name          name of component folder in package
component_job_name      application_id-component_name-job
component_xxx           setting xxx from properties.json
hdfspath_path_name      generated from entries in hdfs.json
````

## Environment Variables ##
These can be obtained with the [environment endpoints API](#environment-endpoints-api)
````
environment_cloudera_manager_host       192.168.1.2
environment_cloudera_manager_password   admin
environment_cloudera_manager_username   admin
environment_cluster_private_key         ./dm.pem
environment_cluster_root_user           cloud-user
environment_hbase_rest_port             20550
environment_hbase_rest_server           cluster-cdh-mgr1
environment_hive_port                   10000
environment_hive_server                 cluster-cdh-mgr1
environment_impala_host                 cluster-cdh-dn0
environment_impala_port                 21050
environment_kafka_brokers               192.168.1.3:9092, ...
environment_kafka_manager               https://192.168.1.4:443
environment_kafka_zookeeper             192.168.1.5:2181, ...
environment_metric_logger_url           hhtp://192.169.1.7:3001/metrics
environment_name_node                   hdfs://cluster-cdh-mgr1:8020
environment_namespace                   platform_app
environment_oozie_uri                   http://cluster-cdh-mgr1:11000/oozie
environment_opentsdb                    192.168.1.6:4242
environment_queue_name                  default
environment_webhdfs_host                cluster-cdh-mgr1
environment_webhdfs_port                50070
environment_yarn_node_managers          cluster-cdh-dn0
environment_yarn_resource_manager_host  cluster-cdh-mgr1
environment_yarn_resource_manager_mr_port 8032
environment_yarn_resource_manager_port  8088
environment_zookeeper_port              2181
environment_zookeeper_quorum            cluster-cdh-mgr1
````

## Oozie Specific Variables ##
The following varibles are only injected for Oozie components.

````
component_end                  2016-03-31T17:07Z
component_start                2016-03-24T17:07Z
mapreduce.job.user.name        hdfs
oozie.coord.application.path   hdfs://cluster-cdh-mgr1:8020/user/application_id/component_name/coordinator.xml
oozie.libpath                  /user/deployment/platform
oozie.use.system.libpath       true
user.name                      hdfs
```` 


CHANGELOG
# Change Log
All notable changes to this project will be documented in this file.

## [0.1.0] 2016-07-01
### First version

## [Pre-release]

### Added

- Add hue endpoint to environment endpoints API
- Application names checked to only contain alphanumeric characters (a-z A-Z 0-9 - and _) because they are used directly in file paths.
- Added ability to discover HDFS namedservices
- Added information field to status reports
- Using an external pacakge repository API instead of internal swift integration
- Application detail API (GET /applications/<application>/detail) now returns YARN IDs assigned to the running tasks for that application.
- Oozie error messages are reported when querying for status of an application creation call.
- Packages are validated on deployment and the error messages reported when querying for status of a package deployment call.
- Added support for opentsdb.json descriptor for creating metrics when deploying applications.
- Callback events are sent to the console data logger.
- Application detail API now completed to return Yarn IDs for any Yarn applications associated with a PNDA application.

### Fixed 

- Return IP address for webhdfs/HTTPFS endpoint instead of hostname
- Timeout calls to package repository at 120 seconds.
- Deploying a package that does not exist in the package repository now results in a useful error message being returned to the caller.
- Fixed defect preventing '-' being used in application names.
- Fix Zookeeper quorum bug issue
- Improve package validation to catch packages without 3 point version numbers and where the folder inside the tar does not match the package name.
- Add list of zookeeper nodes to quorum
- Remove port=8020 for named service
- Oozie creator plugin sets 'oozie.wf.application.path' and 'oozie.coord.application.path' to point at the folder not the xml files.
- Removed some stdout printouts
- Fixed bug preventing recency parameter being used on the repository/packages API.


api/src/main/resources/deployer_utils.py
"""
Name:       deployer_utils.py
Purpose:    Utility functions for general usage in the project
Author:     PNDA team

Created:    21/03/2016

Copyright (c) 2016 Cisco and/or its affiliates.

This software is licensed to you under the terms of the Apache License, Version 2.0 (the "License").
You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0

The code, technical concepts, and all information contained herein, are the property of Cisco Technology, Inc.
and/or its affiliated entities, under various laws including copyright, international treaties, patent,
and/or contract. Any use of the material herein must be in accordance with the terms of the License.
All rights not expressly granted by the License are reserved.

Unless required by applicable law or agreed to separately in writing, software distributed under the
License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
either express or implied.
"""

import os
import tarfile
import StringIO
import logging
import traceback
import spur
from pywebhdfs.webhdfs import PyWebHdfsClient

import requests
from cm_api.api_client import ApiResource


def connect_cm(cm_api, cm_username, cm_password):
    api = ApiResource(
        cm_api,
        version=6,
        username=cm_username,
        password=cm_password)
    return api


def get_named_service(cm_host, cluster_name, service_name, user_name='admin', password='admin'):
    request_url = 'http://%s:7180/api/v11/clusters/%s/services/%s/nameservices' % (cm_host,
                                                                                   cluster_name,
                                                                                   service_name)
    result = requests.get(request_url, auth=(user_name, password))
    named_service = ""
    if result.status_code == 200:
        response = result.json()
        if 'items' in response:
            named_service = response['items'][0]['name']
            logging.debug("Found named service %s for %s", named_service, service_name)
    return named_service


def fill_hadoop_env(env):
    # pylint: disable=E1103
    api = connect_cm(
        env['cloudera_manager_host'],
        env['cloudera_manager_username'],
        env['cloudera_manager_password'])

    for cluster_detail in api.get_all_clusters():
        cluster_name = cluster_detail.name
        break

    logging.debug('getting ' + cluster_name)
    env['cm_status_links'] = {}

    cluster = api.get_cluster(cluster_name)
    for service in cluster.get_all_services():
        env['cm_status_links']['%s' % service.name] = service.serviceUrl
        if service.type == "HDFS":
            named_service = get_named_service(env['cloudera_manager_host'], cluster_name,
                                              service.name,
                                              user_name=env['cloudera_manager_username'],
                                              password=env['cloudera_manager_password'])
            if named_service:
                env['name_node'] = 'hdfs://%s' % named_service
            for role in service.get_all_roles():
                if not named_service and role.type == "NAMENODE":
                    env['name_node'] = 'hdfs://%s:8020' % api.get_host(role.hostRef.hostId).hostname
                if role.type == "HTTPFS":
                    env['webhdfs_host'] = '%s' % api.get_host(role.hostRef.hostId).ipAddress
                    env['webhdfs_port'] = '14000'
        elif service.type == "YARN":
            for role in service.get_all_roles():
                if role.type == "RESOURCEMANAGER":
                    env['yarn_resource_manager_host'] = '%s' % api.get_host(
                        role.hostRef.hostId).hostname
                    env['yarn_resource_manager_port'] = '8088'
                    env['yarn_resource_manager_mr_port'] = '8032'
                if role.type == "NODEMANAGER":
                    if 'yarn_node_managers' in env:
                        env['yarn_node_managers'] = '%s,%s' % (
                            env['yarn_node_managers'], api.get_host(role.hostRef.hostId).hostname)
                    else:
                        env['yarn_node_managers'] = '%s' % api.get_host(
                            role.hostRef.hostId).hostname
        elif service.type == "MAPREDUCE":
            for role in service.get_all_roles():
                if role.type == "JOBTRACKER":
                    env['job_tracker'] = '%s:8021' % api.get_host(
                        role.hostRef.hostId).hostname
                    break
        elif service.type == "ZOOKEEPER":
            for role in service.get_all_roles():
                if role.type == "SERVER":
                    if 'zookeeper_quorum' in env:
                        env['zookeeper_quorum'] += ',%s' % api.get_host(role.hostRef.hostId).hostname
                    else:
                        env['zookeeper_quorum'] = '%s' % api.get_host(role.hostRef.hostId).hostname
                        env['zookeeper_port'] = '2181'
        elif service.type == "HBASE":
            for role in service.get_all_roles():
                if role.type == "HBASERESTSERVER":
                    env['hbase_rest_server'] = '%s' % api.get_host(
                        role.hostRef.hostId).hostname
                    env['hbase_rest_port'] = '20550'
                    break
        elif service.type == "OOZIE":
            for role in service.get_all_roles():
                if role.type == "OOZIE_SERVER":
                    env['oozie_uri'] = 'http://%s:11000/oozie' % api.get_host(
                        role.hostRef.hostId).hostname
                    break
        elif service.type == "HIVE":
            for role in service.get_all_roles():
                if role.type == "HIVESERVER2":
                    env['hive_server'] = '%s' % api.get_host(
                        role.hostRef.hostId).hostname
                    env['hive_port'] = '10000'
                    break
        elif service.type == "IMPALA":
            for role in service.get_all_roles():
                if role.type == "IMPALAD":
                    env['impala_host'] = '%s' % api.get_host(
                        role.hostRef.hostId).hostname
                    env['impala_port'] = '21050'
                    break
        elif service.type == "KUDU":
            for role in service.get_all_roles():
                if role.type == "KUDU_MASTER":
                    env['kudu_host'] = '%s' % api.get_host(
                        role.hostRef.hostId).hostname
                    env['kudu_port'] = '7051'
                    break
        elif service.type == "HUE":
            for role in service.get_all_roles():
                if role.type == "HUE_SERVER":
                    env['hue_host'] = '%s' % api.get_host(
                        role.hostRef.hostId).hostname
                    env['hue_port'] = '8888'
                    break

    logging.debug(env)


def tree(archive_filepath):
    file_handle = file(archive_filepath, 'rb')
    tar_file = tarfile.open(None, 'r', file_handle)
    table = tar_file.getmembers()

    root = {}
    for member in table:
        path = member.name.split('/')
        node = root
        for part in path:
            if part not in node:
                node[part] = {}
            node = node[part]

    return root


def canonicalize(path):
    path = path.replace('\\', '/')
    path = path.replace('//', '/')
    return path


class HDFS(object):
    def __init__(self, host, port, user):
        self._hdfs = PyWebHdfsClient(
            host=host, port=port, user_name=user, timeout=None)
        logging.debug('webhdfs = %s@%s:%s', user, host, port)

    def recursive_copy(self, local_path, remote_path, exclude=None):

        if exclude is None:
            exclude = []

        c_path = canonicalize(remote_path)
        logging.debug('making %s', c_path)
        self._hdfs.make_dir(c_path)

        fs_g = os.walk(local_path)
        for dpath, dnames, fnames in fs_g:
            _, relative_path = dpath.split(local_path)
            for dname in dnames:
                if dname not in exclude:
                    c_path = canonicalize(
                        '%s/%s/%s' %
                        (remote_path, relative_path, dname))
                    logging.debug('making %s', c_path)
                    self._hdfs.make_dir(c_path)

            for fname in fnames:
                if fname not in exclude:
                    data = file(
                        canonicalize(
                            '%s/%s/%s' %
                            (local_path, relative_path, fname)), 'rb')
                    c_path = canonicalize(
                        '%s/%s/%s' %
                        (remote_path, relative_path, fname))
                    logging.debug('creating %s', c_path)
                    self._hdfs.create_file(c_path, data, overwrite=True)
                    data.close()

    def make_dir(self, path):

        logging.debug('make_dir: %s', path)

        self._hdfs.make_dir(canonicalize(path))

    def create_file(self, data, remote_file_path):

        logging.debug('create_file: %s', remote_file_path)

        sio = StringIO.StringIO(data)

        self._hdfs.create_file(
            canonicalize(remote_file_path),
            sio,
            overwrite=True)

    def read_file(self, remote_file_path):

        data = self._hdfs.read_file(canonicalize(remote_file_path))

        return data

    def remove(self, path, recursive=False):

        logging.debug('remove: %s', path)

        self._hdfs.delete_file_dir(canonicalize(path), recursive)


def exec_ssh(host, user, key, ssh_commands):
    shell = spur.SshShell(
        hostname=host,
        username=user,
        private_key_file=key,
        missing_host_key=spur.ssh.MissingHostKey.accept)
    with shell:
        for ssh_command in ssh_commands:
            logging.debug('Host - %s: Command - %s', host, ssh_command)
            try:
                shell.run(["bash", "-c", ssh_command])
            except spur.results.RunProcessError as exception:
                logging.error(
                    ssh_command +
                    " - error: " +
                    traceback.format_exc(exception))


def dict_to_props(dict_props):
    props = []
    for key, value in dict_props.iteritems():
        props.append('%s=%s' % (key, value))
    return '\n'.join(props)


def dict_to_xml(dict_props):
    xml_header = '<?xml version="1.0" encoding="UTF-8" ?>'
    xml_string = '<configuration>'
    for key, value in dict_props.items():
        xml_string += '<property>' + \
                      '<name>' + key + '</name>' + \
                      '<value>' + str(value) + '</value>' + \
                      '</property>'
    xml_string += '</configuration>'
    return xml_header + xml_string
